docs/SsoConfiguration.md
# ConnectWise::SsoConfiguration

## Properties

| Name | Type | Description | Notes |
| ---- | ---- | ----------- | ----- |
| **name** | **String** | Descriptor of the SSO Configuration Max length: 100; |  |
| **sso_type** | **String** | Type of SSO Configuration |  |
| **location_ids** | **Array&lt;Integer&gt;** | The locations where the SAML Idp Configuration is used |  |
| **id** | **Integer** | Unique identifier of the SSO Configuration | [optional] |
| **inactive_flag** | **Boolean** | Whether the SSO configuration is not active | [optional] |
| **saml_entity_id** | **String** | SAML Identity Provider Id Max length: 1000; | [optional] |
| **saml_sign_in_url** | **String** | Sign in url for the SAML Identity Provider Max length: 1000; | [optional] |
| **saml_idp_certificate** | **String** | Public certificate for Identity Provider signatures | [optional] |
| **saml_certificate_name** | **String** | Name of the SAML certificate. Metadata on SAML_Idp_Certificate | [optional] |
| **saml_certificate_issued_to** | **String** | Who the SAML certificate was issued to. Metadata on SAML_Idp_Certificate | [optional] |
| **saml_certificate_thumbprint** | **String** | Thumbprint of the SAML certificate. Metadata on SAML_Idp_Certificate | [optional] |
| **saml_certificate_valid_from** | **Time** | Date when the SAML certificate becomes valid. Metadata on SAML_Idp_Certificate | [optional] |
| **saml_certificate_valid_to** | **Time** | Date when the SAML certificate is no longer valid. Metadata on SAML_Idp_Certificate | [optional] |
| **client_id** | **String** | Client identity for this configuration of ConnectWise SSO Max length: 1000; | [optional] |
| **sts_base_url** | **String** | Sign in URL for ConnectWise SSO | [optional] |
| **sts_user_admin_url** | **String** | User Admin Url for ConnectWise SSO | [optional] |
| **token** | **String** |  | [optional] |
| **submitted_member_count** | **Integer** |  | [optional] |
| **all_members_submitted** | **Boolean** |  | [optional] |
| **_info** | **Hash&lt;String, String&gt;** |  | [optional] |
| **is_sso_on_by_default** | **Boolean** |  | [optional] |

## Example

```ruby
require 'connect_wise'

instance = ConnectWise::SsoConfiguration.new(
  name: null,
  sso_type: null,
  location_ids: null,
  id: null,
  inactive_flag: null,
  saml_entity_id: null,
  saml_sign_in_url: null,
  saml_idp_certificate: null,
  saml_certificate_name: null,
  saml_certificate_issued_to: null,
  saml_certificate_thumbprint: null,
  saml_certificate_valid_from: null,
  saml_certificate_valid_to: null,
  client_id: null,
  sts_base_url: null,
  sts_user_admin_url: null,
  token: null,
  submitted_member_count: null,
  all_members_submitted: null,
  _info: null,
  is_sso_on_by_default: null
)
```



docs/SurveyOptionsApi.md
# ConnectWise::SurveyOptionsApi

All URIs are relative to *http://cloud.na.myconnectwise.net/v4_6_development/apis/3.0*

| Method | HTTP request | Description |
| ------ | ------------ | ----------- |
| [**delete_service_surveys_by_grandparent_id_questions_by_parent_id_options_by_id**](SurveyOptionsApi.md#delete_service_surveys_by_grandparent_id_questions_by_parent_id_options_by_id) | **DELETE** /service/surveys/{grandparentId}/questions/{parentId}/options/{id} | Delete SurveyOption |
| [**get_service_surveys_by_grandparent_id_questions_by_parent_id_options**](SurveyOptionsApi.md#get_service_surveys_by_grandparent_id_questions_by_parent_id_options) | **GET** /service/surveys/{grandparentId}/questions/{parentId}/options | Get List of SurveyOption |
| [**get_service_surveys_by_grandparent_id_questions_by_parent_id_options_by_id**](SurveyOptionsApi.md#get_service_surveys_by_grandparent_id_questions_by_parent_id_options_by_id) | **GET** /service/surveys/{grandparentId}/questions/{parentId}/options/{id} | Get SurveyOption |
| [**get_service_surveys_by_grandparent_id_questions_by_parent_id_options_count**](SurveyOptionsApi.md#get_service_surveys_by_grandparent_id_questions_by_parent_id_options_count) | **GET** /service/surveys/{grandparentId}/questions/{parentId}/options/count | Get Count of SurveyOption |
| [**patch_service_surveys_by_grandparent_id_questions_by_parent_id_options_by_id**](SurveyOptionsApi.md#patch_service_surveys_by_grandparent_id_questions_by_parent_id_options_by_id) | **PATCH** /service/surveys/{grandparentId}/questions/{parentId}/options/{id} | Patch SurveyOption |
| [**post_service_surveys_by_grandparent_id_questions_by_parent_id_options**](SurveyOptionsApi.md#post_service_surveys_by_grandparent_id_questions_by_parent_id_options) | **POST** /service/surveys/{grandparentId}/questions/{parentId}/options | Post SurveyOption |
| [**put_service_surveys_by_grandparent_id_questions_by_parent_id_options_by_id**](SurveyOptionsApi.md#put_service_surveys_by_grandparent_id_questions_by_parent_id_options_by_id) | **PUT** /service/surveys/{grandparentId}/questions/{parentId}/options/{id} | Put SurveyOption |


## delete_service_surveys_by_grandparent_id_questions_by_parent_id_options_by_id

> delete_service_surveys_by_grandparent_id_questions_by_parent_id_options_by_id(id, parent_id, grandparent_id, client_id)

Delete SurveyOption

### Examples

```ruby
require 'time'
require 'connect_wise'

api_instance = ConnectWise::SurveyOptionsApi.new
id = 56 # Integer | optionId
parent_id = 56 # Integer | questionId
grandparent_id = 56 # Integer | surveyId
client_id = 'client_id_example' # String | 

begin
  # Delete SurveyOption
  api_instance.delete_service_surveys_by_grandparent_id_questions_by_parent_id_options_by_id(id, parent_id, grandparent_id, client_id)
rescue ConnectWise::ApiError => e
  puts "Error when calling SurveyOptionsApi->delete_service_surveys_by_grandparent_id_questions_by_parent_id_options_by_id: #{e}"
end
```

#### Using the delete_service_surveys_by_grandparent_id_questions_by_parent_id_options_by_id_with_http_info variant

This returns an Array which contains the response data (`nil` in this case), status code and headers.

> <Array(nil, Integer, Hash)> delete_service_surveys_by_grandparent_id_questions_by_parent_id_options_by_id_with_http_info(id, parent_id, grandparent_id, client_id)

```ruby
begin
  # Delete SurveyOption
  data, status_code, headers = api_instance.delete_service_surveys_by_grandparent_id_questions_by_parent_id_options_by_id_with_http_info(id, parent_id, grandparent_id, client_id)
  p status_code # => 2xx
  p headers # => { ... }
  p data # => nil
rescue ConnectWise::ApiError => e
  puts "Error when calling SurveyOptionsApi->delete_service_surveys_by_grandparent_id_questions_by_parent_id_options_by_id_with_http_info: #{e}"
end
```

### Parameters

| Name | Type | Description | Notes |
| ---- | ---- | ----------- | ----- |
| **id** | **Integer** | optionId |  |
| **parent_id** | **Integer** | questionId |  |
| **grandparent_id** | **Integer** | surveyId |  |
| **client_id** | **String** |  |  |

### Return type

nil (empty response body)

### Authorization

No authorization required

### HTTP request headers

- **Content-Type**: Not defined
- **Accept**: Not defined


## get_service_surveys_by_grandparent_id_questions_by_parent_id_options

> <Array<SurveyOption>> get_service_surveys_by_grandparent_id_questions_by_parent_id_options(parent_id, grandparent_id, client_id, opts)

Get List of SurveyOption

### Examples

```ruby
require 'time'
require 'connect_wise'

api_instance = ConnectWise::SurveyOptionsApi.new
parent_id = 56 # Integer | questionId
grandparent_id = 56 # Integer | surveyId
client_id = 'client_id_example' # String | 
opts = {
  conditions: 'conditions_example', # String | 
  child_conditions: 'child_conditions_example', # String | 
  custom_field_conditions: 'custom_field_conditions_example', # String | 
  order_by: 'order_by_example', # String | 
  fields: 'fields_example', # String | 
  page: 56, # Integer | 
  page_size: 56, # Integer | 
  page_id: 56 # Integer | 
}

begin
  # Get List of SurveyOption
  result = api_instance.get_service_surveys_by_grandparent_id_questions_by_parent_id_options(parent_id, grandparent_id, client_id, opts)
  p result
rescue ConnectWise::ApiError => e
  puts "Error when calling SurveyOptionsApi->get_service_surveys_by_grandparent_id_questions_by_parent_id_options: #{e}"
end
```

#### Using the get_service_surveys_by_grandparent_id_questions_by_parent_id_options_with_http_info variant

This returns an Array which contains the response data, status code and headers.

> <Array(<Array<SurveyOption>>, Integer, Hash)> get_service_surveys_by_grandparent_id_questions_by_parent_id_options_with_http_info(parent_id, grandparent_id, client_id, opts)

```ruby
begin
  # Get List of SurveyOption
  data, status_code, headers = api_instance.get_service_surveys_by_grandparent_id_questions_by_parent_id_options_with_http_info(parent_id, grandparent_id, client_id, opts)
  p status_code # => 2xx
  p headers # => { ... }
  p data # => <Array<SurveyOption>>
rescue ConnectWise::ApiError => e
  puts "Error when calling SurveyOptionsApi->get_service_surveys_by_grandparent_id_questions_by_parent_id_options_with_http_info: #{e}"
end
```

### Parameters

| Name | Type | Description | Notes |
| ---- | ---- | ----------- | ----- |
| **parent_id** | **Integer** | questionId |  |
| **grandparent_id** | **Integer** | surveyId |  |
| **client_id** | **String** |  |  |
| **conditions** | **String** |  | [optional] |
| **child_conditions** | **String** |  | [optional] |
| **custom_field_conditions** | **String** |  | [optional] |
| **order_by** | **String** |  | [optional] |
| **fields** | **String** |  | [optional] |
| **page** | **Integer** |  | [optional] |
| **page_size** | **Integer** |  | [optional] |
| **page_id** | **Integer** |  | [optional] |

### Return type

[**Array&lt;SurveyOption&gt;**](SurveyOption.md)

### Authorization

No authorization required

### HTTP request headers

- **Content-Type**: Not defined
- **Accept**: application/vnd.connectwise.com+json; version=2022.1


## get_service_surveys_by_grandparent_id_questions_by_parent_id_options_by_id

> <SurveyOption> get_service_surveys_by_grandparent_id_questions_by_parent_id_options_by_id(id, parent_id, grandparent_id, client_id, opts)

Get SurveyOption

### Examples

```ruby
require 'time'
require 'connect_wise'

api_instance = ConnectWise::SurveyOptionsApi.new
id = 56 # Integer | optionId
parent_id = 56 # Integer | questionId
grandparent_id = 56 # Integer | surveyId
client_id = 'client_id_example' # String | 
opts = {
  conditions: 'conditions_example', # String | 
  child_conditions: 'child_conditions_example', # String | 
  custom_field_conditions: 'custom_field_conditions_example', # String | 
  order_by: 'order_by_example', # String | 
  fields: 'fields_example', # String | 
  page: 56, # Integer | 
  page_size: 56, # Integer | 
  page_id: 56 # Integer | 
}

begin
  # Get SurveyOption
  result = api_instance.get_service_surveys_by_grandparent_id_questions_by_parent_id_options_by_id(id, parent_id, grandparent_id, client_id, opts)
  p result
rescue ConnectWise::ApiError => e
  puts "Error when calling SurveyOptionsApi->get_service_surveys_by_grandparent_id_questions_by_parent_id_options_by_id: #{e}"
end
```

#### Using the get_service_surveys_by_grandparent_id_questions_by_parent_id_options_by_id_with_http_info variant

This returns an Array which contains the response data, status code and headers.

> <Array(<SurveyOption>, Integer, Hash)> get_service_surveys_by_grandparent_id_questions_by_parent_id_options_by_id_with_http_info(id, parent_id, grandparent_id, client_id, opts)

```ruby
begin
  # Get SurveyOption
  data, status_code, headers = api_instance.get_service_surveys_by_grandparent_id_questions_by_parent_id_options_by_id_with_http_info(id, parent_id, grandparent_id, client_id, opts)
  p status_code # => 2xx
  p headers # => { ... }
  p data # => <SurveyOption>
rescue ConnectWise::ApiError => e
  puts "Error when calling SurveyOptionsApi->get_service_surveys_by_grandparent_id_questions_by_parent_id_options_by_id_with_http_info: #{e}"
end
```

### Parameters

| Name | Type | Description | Notes |
| ---- | ---- | ----------- | ----- |
| **id** | **Integer** | optionId |  |
| **parent_id** | **Integer** | questionId |  |
| **grandparent_id** | **Integer** | surveyId |  |
| **client_id** | **String** |  |  |
| **conditions** | **String** |  | [optional] |
| **child_conditions** | **String** |  | [optional] |
| **custom_field_conditions** | **String** |  | [optional] |
| **order_by** | **String** |  | [optional] |
| **fields** | **String** |  | [optional] |
| **page** | **Integer** |  | [optional] |
| **page_size** | **Integer** |  | [optional] |
| **page_id** | **Integer** |  | [optional] |

### Return type

[**SurveyOption**](SurveyOption.md)

### Authorization

No authorization required

### HTTP request headers

- **Content-Type**: Not defined
- **Accept**: application/vnd.connectwise.com+json; version=2022.1


## get_service_surveys_by_grandparent_id_questions_by_parent_id_options_count

> <Count> get_service_surveys_by_grandparent_id_questions_by_parent_id_options_count(parent_id, grandparent_id, client_id, opts)

Get Count of SurveyOption

### Examples

```ruby
require 'time'
require 'connect_wise'

api_instance = ConnectWise::SurveyOptionsApi.new
parent_id = 56 # Integer | questionId
grandparent_id = 56 # Integer | surveyId
client_id = 'client_id_example' # String | 
opts = {
  conditions: 'conditions_example', # String | 
  child_conditions: 'child_conditions_example', # String | 
  custom_field_conditions: 'custom_field_conditions_example', # String | 
  order_by: 'order_by_example', # String | 
  fields: 'fields_example', # String | 
  page: 56, # Integer | 
  page_size: 56, # Integer | 
  page_id: 56 # Integer | 
}

begin
  # Get Count of SurveyOption
  result = api_instance.get_service_surveys_by_grandparent_id_questions_by_parent_id_options_count(parent_id, grandparent_id, client_id, opts)
  p result
rescue ConnectWise::ApiError => e
  puts "Error when calling SurveyOptionsApi->get_service_surveys_by_grandparent_id_questions_by_parent_id_options_count: #{e}"
end
```

#### Using the get_service_surveys_by_grandparent_id_questions_by_parent_id_options_count_with_http_info variant

This returns an Array which contains the response data, status code and headers.

> <Array(<Count>, Integer, Hash)> get_service_surveys_by_grandparent_id_questions_by_parent_id_options_count_with_http_info(parent_id, grandparent_id, client_id, opts)

```ruby
begin
  # Get Count of SurveyOption
  data, status_code, headers = api_instance.get_service_surveys_by_grandparent_id_questions_by_parent_id_options_count_with_http_info(parent_id, grandparent_id, client_id, opts)
  p status_code # => 2xx
  p headers # => { ... }
  p data # => <Count>
rescue ConnectWise::ApiError => e
  puts "Error when calling SurveyOptionsApi->get_service_surveys_by_grandparent_id_questions_by_parent_id_options_count_with_http_info: #{e}"
end
```

### Parameters

| Name | Type | Description | Notes |
| ---- | ---- | ----------- | ----- |
| **parent_id** | **Integer** | questionId |  |
| **grandparent_id** | **Integer** | surveyId |  |
| **client_id** | **String** |  |  |
| **conditions** | **String** |  | [optional] |
| **child_conditions** | **String** |  | [optional] |
| **custom_field_conditions** | **String** |  | [optional] |
| **order_by** | **String** |  | [optional] |
| **fields** | **String** |  | [optional] |
| **page** | **Integer** |  | [optional] |
| **page_size** | **Integer** |  | [optional] |
| **page_id** | **Integer** |  | [optional] |

### Return type

[**Count**](Count.md)

### Authorization

No authorization required

### HTTP request headers

- **Content-Type**: Not defined
- **Accept**: application/vnd.connectwise.com+json; version=2022.1


## patch_service_surveys_by_grandparent_id_questions_by_parent_id_options_by_id

> <SurveyOption> patch_service_surveys_by_grandparent_id_questions_by_parent_id_options_by_id(id, parent_id, grandparent_id, client_id, patch_operation)

Patch SurveyOption

### Examples

```ruby
require 'time'
require 'connect_wise'

api_instance = ConnectWise::SurveyOptionsApi.new
id = 56 # Integer | optionId
parent_id = 56 # Integer | questionId
grandparent_id = 56 # Integer | surveyId
client_id = 'client_id_example' # String | 
patch_operation = [ConnectWise::PatchOperation.new] # Array<PatchOperation> | List of PatchOperation

begin
  # Patch SurveyOption
  result = api_instance.patch_service_surveys_by_grandparent_id_questions_by_parent_id_options_by_id(id, parent_id, grandparent_id, client_id, patch_operation)
  p result
rescue ConnectWise::ApiError => e
  puts "Error when calling SurveyOptionsApi->patch_service_surveys_by_grandparent_id_questions_by_parent_id_options_by_id: #{e}"
end
```

#### Using the patch_service_surveys_by_grandparent_id_questions_by_parent_id_options_by_id_with_http_info variant

This returns an Array which contains the response data, status code and headers.

> <Array(<SurveyOption>, Integer, Hash)> patch_service_surveys_by_grandparent_id_questions_by_parent_id_options_by_id_with_http_info(id, parent_id, grandparent_id, client_id, patch_operation)

```ruby
begin
  # Patch SurveyOption
  data, status_code, headers = api_instance.patch_service_surveys_by_grandparent_id_questions_by_parent_id_options_by_id_with_http_info(id, parent_id, grandparent_id, client_id, patch_operation)
  p status_code # => 2xx
  p headers # => { ... }
  p data # => <SurveyOption>
rescue ConnectWise::ApiError => e
  puts "Error when calling SurveyOptionsApi->patch_service_surveys_by_grandparent_id_questions_by_parent_id_options_by_id_with_http_info: #{e}"
end
```

### Parameters

| Name | Type | Description | Notes |
| ---- | ---- | ----------- | ----- |
| **id** | **Integer** | optionId |  |
| **parent_id** | **Integer** | questionId |  |
| **grandparent_id** | **Integer** | surveyId |  |
| **client_id** | **String** |  |  |
| **patch_operation** | [**Array&lt;PatchOperation&gt;**](PatchOperation.md) | List of PatchOperation |  |

### Return type

[**SurveyOption**](SurveyOption.md)

### Authorization

No authorization required

### HTTP request headers

- **Content-Type**: application/json
- **Accept**: application/vnd.connectwise.com+json; version=2022.1


## post_service_surveys_by_grandparent_id_questions_by_parent_id_options

> <SurveyOption> post_service_surveys_by_grandparent_id_questions_by_parent_id_options(parent_id, grandparent_id, client_id, survey_option)

Post SurveyOption

### Examples

```ruby
require 'time'
require 'connect_wise'

api_instance = ConnectWise::SurveyOptionsApi.new
parent_id = 56 # Integer | questionId
grandparent_id = 56 # Integer | surveyId
client_id = 'client_id_example' # String | 
survey_option = ConnectWise::SurveyOption.new({caption: 'caption_example', points: 37}) # SurveyOption | surveyOption

begin
  # Post SurveyOption
  result = api_instance.post_service_surveys_by_grandparent_id_questions_by_parent_id_options(parent_id, grandparent_id, client_id, survey_option)
  p result
rescue ConnectWise::ApiError => e
  puts "Error when calling SurveyOptionsApi->post_service_surveys_by_grandparent_id_questions_by_parent_id_options: #{e}"
end
```

#### Using the post_service_surveys_by_grandparent_id_questions_by_parent_id_options_with_http_info variant

This returns an Array which contains the response data, status code and headers.

> <Array(<SurveyOption>, Integer, Hash)> post_service_surveys_by_grandparent_id_questions_by_parent_id_options_with_http_info(parent_id, grandparent_id, client_id, survey_option)

```ruby
begin
  # Post SurveyOption
  data, status_code, headers = api_instance.post_service_surveys_by_grandparent_id_questions_by_parent_id_options_with_http_info(parent_id, grandparent_id, client_id, survey_option)
  p status_code # => 2xx
  p headers # => { ... }
  p data # => <SurveyOption>
rescue ConnectWise::ApiError => e
  puts "Error when calling SurveyOptionsApi->post_service_surveys_by_grandparent_id_questions_by_parent_id_options_with_http_info: #{e}"
end
```

### Parameters

| Name | Type | Description | Notes |
| ---- | ---- | ----------- | ----- |
| **parent_id** | **Integer** | questionId |  |
| **grandparent_id** | **Integer** | surveyId |  |
| **client_id** | **String** |  |  |
| **survey_option** | [**SurveyOption**](SurveyOption.md) | surveyOption |  |

### Return type

[**SurveyOption**](SurveyOption.md)

### Authorization

No authorization required

### HTTP request headers

- **Content-Type**: application/json
- **Accept**: application/vnd.connectwise.com+json; version=2022.1


## put_service_surveys_by_grandparent_id_questions_by_parent_id_options_by_id

> <SurveyOption> put_service_surveys_by_grandparent_id_questions_by_parent_id_options_by_id(id, parent_id, grandparent_id, client_id, survey_option)

Put SurveyOption

### Examples

```ruby
require 'time'
require 'connect_wise'

api_instance = ConnectWise::SurveyOptionsApi.new
id = 56 # Integer | optionId
parent_id = 56 # Integer | questionId
grandparent_id = 56 # Integer | surveyId
client_id = 'client_id_example' # String | 
survey_option = ConnectWise::SurveyOption.new({caption: 'caption_example', points: 37}) # SurveyOption | surveyOption

begin
  # Put SurveyOption
  result = api_instance.put_service_surveys_by_grandparent_id_questions_by_parent_id_options_by_id(id, parent_id, grandparent_id, client_id, survey_option)
  p result
rescue ConnectWise::ApiError => e
  puts "Error when calling SurveyOptionsApi->put_service_surveys_by_grandparent_id_questions_by_parent_id_options_by_id: #{e}"
end
```

#### Using the put_service_surveys_by_grandparent_id_questions_by_parent_id_options_by_id_with_http_info variant

This returns an Array which contains the response data, status code and headers.

> <Array(<SurveyOption>, Integer, Hash)> put_service_surveys_by_grandparent_id_questions_by_parent_id_options_by_id_with_http_info(id, parent_id, grandparent_id, client_id, survey_option)

```ruby
begin
  # Put SurveyOption
  data, status_code, headers = api_instance.put_service_surveys_by_grandparent_id_questions_by_parent_id_options_by_id_with_http_info(id, parent_id, grandparent_id, client_id, survey_option)
  p status_code # => 2xx
  p headers # => { ... }
  p data # => <SurveyOption>
rescue ConnectWise::ApiError => e
  puts "Error when calling SurveyOptionsApi->put_service_surveys_by_grandparent_id_questions_by_parent_id_options_by_id_with_http_info: #{e}"
end
```

### Parameters

| Name | Type | Description | Notes |
| ---- | ---- | ----------- | ----- |
| **id** | **Integer** | optionId |  |
| **parent_id** | **Integer** | questionId |  |
| **grandparent_id** | **Integer** | surveyId |  |
| **client_id** | **String** |  |  |
| **survey_option** | [**SurveyOption**](SurveyOption.md) | surveyOption |  |

### Return type

[**SurveyOption**](SurveyOption.md)

### Authorization

No authorization required

### HTTP request headers

- **Content-Type**: application/json
- **Accept**: application/vnd.connectwise.com+json; version=2022.1



lib/connect_wise/configuration.rb
# #Connectwise Manage All Endpoints
#
# No description provided (generated by Openapi Generator https://github.com/openapitools/openapi-generator)
#
# The version of the OpenAPI document: 2022.1
#
# Generated by: https://openapi-generator.tech
# OpenAPI Generator version: 6.2.1
#

module ConnectWise
  class Configuration
    # Defines url scheme
    attr_accessor :scheme

    # Defines url host
    attr_accessor :host

    # Defines url base path
    attr_accessor :base_path

    # Define server operation configuration index
    attr_accessor :server_operation_index

    # Default server variables
    attr_accessor :server_variables

    # Default server operation variables
    attr_accessor :server_operation_variables

    # Defines API keys used with API Key authentications.
    #
    # @return [Hash] key: parameter name, value: parameter value (API key)
    #
    # @example parameter name is "api_key", API key is "xxx" (e.g. "api_key=xxx" in query string)
    #   config.api_key['api_key'] = 'xxx'
    attr_accessor :api_key

    # Defines API key prefixes used with API Key authentications.
    #
    # @return [Hash] key: parameter name, value: API key prefix
    #
    # @example parameter name is "Authorization", API key prefix is "Token" (e.g. "Authorization: Token xxx" in headers)
    #   config.api_key_prefix['api_key'] = 'Token'
    attr_accessor :api_key_prefix

    # Defines the username used with HTTP basic authentication.
    #
    # @return [String]
    attr_accessor :username

    # Defines the password used with HTTP basic authentication.
    #
    # @return [String]
    attr_accessor :password

    # Defines the access token (Bearer) used with OAuth2.
    attr_accessor :access_token

    # Set this to enable/disable debugging. When enabled (set to true), HTTP request/response
    # details will be logged with `logger.debug` (see the `logger` attribute).
    # Default to false.
    #
    # @return [true, false]
    attr_accessor :debugging

    # Defines the logger used for debugging.
    # Default to `Rails.logger` (when in Rails) or logging to STDOUT.
    #
    # @return [#debug]
    attr_accessor :logger

    # Defines the temporary folder to store downloaded files
    # (for API endpoints that have file response).
    # Default to use `Tempfile`.
    #
    # @return [String]
    attr_accessor :temp_folder_path

    # The time limit for HTTP request in seconds.
    # Default to 0 (never times out).
    attr_accessor :timeout

    # Set this to false to skip client side validation in the operation.
    # Default to true.
    # @return [true, false]
    attr_accessor :client_side_validation

    ### TLS/SSL setting
    # Set this to false to skip verifying SSL certificate when calling API from https server.
    # Default to true.
    #
    # @note Do NOT set it to false in production code, otherwise you would face multiple types of cryptographic attacks.
    #
    # @return [true, false]
    attr_accessor :ssl_verify

    ### TLS/SSL setting
    # Any `OpenSSL::SSL::` constant (see https://ruby-doc.org/stdlib-2.5.1/libdoc/openssl/rdoc/OpenSSL/SSL.html)
    #
    # @note Do NOT set it to false in production code, otherwise you would face multiple types of cryptographic attacks.
    #
    attr_accessor :ssl_verify_mode

    ### TLS/SSL setting
    # Set this to customize the certificate file to verify the peer.
    #
    # @return [String] the path to the certificate file
    attr_accessor :ssl_ca_file

    ### TLS/SSL setting
    # Client certificate file (for client certificate)
    attr_accessor :ssl_client_cert

    ### TLS/SSL setting
    # Client private key file (for client certificate)
    attr_accessor :ssl_client_key

    ### Proxy setting
    # HTTP Proxy settings
    attr_accessor :proxy

    # Set this to customize parameters encoder of array parameter.
    # Default to nil. Faraday uses NestedParamsEncoder when nil.
    #
    # @see The params_encoder option of Faraday. Related source code:
    # https://github.com/lostisland/faraday/tree/main/lib/faraday/encoders
    attr_accessor :params_encoder

    attr_accessor :inject_format, :force_ending_format

    def initialize
      @scheme = 'http'
      @host = 'cloud.na.myconnectwise.net'
      @base_path = '/v4_6_development/apis/3.0'
      @server_operation_index = {}
      @server_variables = {}
      @server_operation_variables = {}
      @api_key = {}
      @api_key_prefix = {}
      @client_side_validation = true
      @ssl_verify = true
      @ssl_verify_mode = nil
      @ssl_ca_file = nil
      @ssl_client_cert = nil
      @ssl_client_key = nil
      @middlewares = Hash.new { |h, k| h[k] = [] }
      @timeout = 60
      # return data as binary instead of file
      @return_binary_data = false
      @params_encoder = nil
      @debugging = false
      @inject_format = false
      @force_ending_format = false
      @logger = defined?(Rails) ? Rails.logger : Logger.new(STDOUT)

      yield(self) if block_given?
    end

    # The default Configuration object.
    def self.default
      @@default ||= Configuration.new
    end

    def configure
      yield(self) if block_given?
    end

    def scheme=(scheme)
      # remove :// from scheme
      @scheme = scheme.sub(%r{://}, '')
    end

    def host=(host)
      # remove http(s):// and anything after a slash
      @host = host.sub(%r{https?://}, '').split('/').first
    end

    def base_path=(base_path)
      # Add leading and trailing slashes to base_path
      @base_path = "/#{base_path}".squeeze('/')
      @base_path = '' if @base_path == '/'
    end

    # Returns base URL for specified operation based on server settings
    def base_url(_operation = nil)
      "#{scheme}://#{[host, base_path].join('/').squeeze('/')}".sub(%r{/+\z}, '')
    end

    # Gets API key (with prefix if set).
    # @param [String] param_name the parameter name of API key auth
    def api_key_with_prefix(param_name, param_alias = nil)
      key = @api_key[param_name]
      key = @api_key.fetch(param_alias, key) unless param_alias.nil?
      if @api_key_prefix[param_name]
        "#{@api_key_prefix[param_name]} #{key}"
      else
        key
      end
    end

    # Gets Basic Auth token string
    def basic_auth_token
      'Basic ' + ["#{username}:#{password}"].pack('m').delete("\r\n")
    end

    # Returns Auth Settings hash for api client.
    def auth_settings
      {
        'basicAuth' =>
          {
            type: 'basic',
            in: 'header',
            key: 'Authorization',
            value: basic_auth_token
          }
      }
    end

    # Returns an array of Server setting
    def server_settings
      [
        {
          url: 'http://cloud.na.myconnectwise.net/v4_6_development/apis/3.0',
          description: 'No description provided'
        }
      ]
    end

    def operation_server_settings
      {}
    end

    # Returns URL based on server settings
    #
    # @param index array index of the server settings
    # @param variables hash of variable and the corresponding value
    def server_url(index, variables = {}, servers = nil)
      servers = server_settings if servers.nil?

      # check array index out of bound
      raise ArgumentError, "Invalid index #{index} when selecting the server. Must be less than #{servers.size}" if index < 0 || index >= servers.size

      server = servers[index]
      url = server[:url]

      return url unless server.key? :variables

      # go through variable and assign a value
      server[:variables].each do |name, _variable|
        if variables.key?(name)
          raise ArgumentError, "The variable `#{name}` in the server URL has invalid value #{variables[name]}. Must be #{server[:variables][name][:enum_values]}." unless !server[:variables][name].key?(:enum_values) || server[:variables][name][:enum_values].include?(variables[name])

          url.gsub! '{' + name.to_s + '}', variables[name]

        else
          # use default value
          url.gsub! '{' + name.to_s + '}', server[:variables][name][:default_value]
        end
      end

      url
    end

    # Adds middleware to the stack
    def use(*middleware)
      set_faraday_middleware(:use, *middleware)
    end

    # Adds request middleware to the stack
    def request(*middleware)
      set_faraday_middleware(:request, *middleware)
    end

    # Adds response middleware to the stack
    def response(*middleware)
      set_faraday_middleware(:response, *middleware)
    end

    # Adds Faraday middleware setting information to the stack
    #
    # @example Use the `set_faraday_middleware` method to set middleware information
    #   config.set_faraday_middleware(:request, :retry, max: 3, methods: [:get, :post], retry_statuses: [503])
    #   config.set_faraday_middleware(:response, :logger, nil, { bodies: true, log_level: :debug })
    #   config.set_faraday_middleware(:use, Faraday::HttpCache, store: Rails.cache, shared_cache: false)
    #   config.set_faraday_middleware(:insert, 0, FaradayMiddleware::FollowRedirects, { standards_compliant: true, limit: 1 })
    #   config.set_faraday_middleware(:swap, 0, Faraday::Response::Logger)
    #   config.set_faraday_middleware(:delete, Faraday::Multipart::Middleware)
    #
    # @see https://github.com/lostisland/faraday/blob/v2.3.0/lib/faraday/rack_builder.rb#L92-L143
    def set_faraday_middleware(operation, key, *args, &block)
      unless %i[request response use insert insert_before insert_after swap delete].include?(operation)
        raise ArgumentError, "Invalid faraday middleware operation #{operation}. Must be" \
                             ' :request, :response, :use, :insert, :insert_before, :insert_after, :swap or :delete.'
      end

      @middlewares[operation] << [key, args, block]
    end
    ruby2_keywords(:set_faraday_middleware) if respond_to?(:ruby2_keywords, true)

    # Set up middleware on the connection
    def configure_middleware(connection)
      return if @middlewares.empty?

      %i[request response use insert insert_before insert_after swap].each do |operation|
        next unless @middlewares.key?(operation)

        @middlewares[operation].each do |key, args, block|
          connection.builder.send(operation, key, *args, &block)
        end
      end

      return unless @middlewares.key?(:delete)

      @middlewares[:delete].each do |key, _args, _block|
        connection.builder.delete(key)
      end
    end
  end
end
